conan/tools/microsoft/nmaketoolchain.py

```diff
@@ -15,42 +16,122 @@ def __init__(self, conanfile):
 
     @property
+    def _ldflags(self):
+        bt_ldflags = self._build_type_link_flags if self._build_type_link_flags else []
+        conf_shared_ldflags = self._conanfile.conf.get("tools.build:sharedlinkflags", default=[], check_type=list)
+        conf_exe_ldflags = self._conanfile.conf.get("tools.build:exelinkflags", default=[], check_type=list)
+        return self._sanitized_options(bt_ldflags + conf_shared_ldflags + conf_exe_ldflags + self.extra_ldflags)
+
+    @property
+    def _defines(self):
+        ndebug_defines = [self._ndebug] if self._ndebug else []
+        conf_defines = self._conanfile.conf.get("tools.build:defines", default=[], check_type=list)
+        return self._sanitized_defines(ndebug_defines + conf_defines + self.extra_defines)
+
+    @property
+    def _cl(self):
+        nologo = ["/nologo"]
+        conf_cflags = self._conanfile.conf.get("tools.build:cflags", default=[], check_type=list)
+        return nologo + self._cxxflags + self._sanitized_options(conf_cflags) + self._defines
+
+    @property
+    def _link(self):
+        nologo = ["/nologo"]
+        return nologo + self._ldflags
+
     def environment(self):
-        # TODO: Seems we want to make this uniform, equal to other generators
-        if self._environment is None:
-            env = Environment()
-            # The whole injection of toolchain happens in CL env-var, the others LIBS, _LINK_
-            env.append("CL", self.cl_flags)
-            self._environment = env
-        return self._environment
-
-    def vars(self, scope="build"):
-        return self.environment.vars(self._conanfile, scope=scope)
-
-    def generate(self, scope="build"):
-        self.vars(scope).save_script("conannmaketoolchain")
-        from conan.tools.microsoft import VCVars
-        VCVars(self._conanfile).generate()
+        env = Environment()
+        # Injection of compile flags in CL env-var:
+        # https://learn.microsoft.com/en-us/cpp/build/reference/cl-environment-variables
+        env.append("CL", self._cl)
+        # Injection of link flags in _LINK_ env-var:
+        # https://learn.microsoft.com/en-us/cpp/build/reference/linking
+        env.append("_LINK_", self._link)
+        # Also define some special env-vars which can override special NMake macros:
+        # https://learn.microsoft.com/en-us/cpp/build/reference/special-nmake-macros
+        conf_compilers = self._conanfile.conf.get("tools.build:compiler_executables", default={}, check_type=dict)
+        if conf_compilers:
+            compilers_mapping = {
+                "AS": "asm",
+                "CC": "c",
+                "CPP": "cpp",
+                "CXX": "cpp",
+                "RC": "rc",
+            }
+            for env_var, comp in compilers_mapping.items():
+                if comp in conf_compilers:
+                    env.define(env_var, conf_compilers[comp])
```

Comment:
Out of scope of this PR, but I believe this toolchain should set all these env vars by default for clang-cl if their corresponding build:compiler_executables is not set.

It would be nice to avoid this kind of fragile logic
https://github.com/conan-io/conan-center-index/blob/2ecc63a88002b1294f0ea0650f27c6012933aafa/recipes/libjpeg/all/conanfile.py#L33-L35
https://github.com/conan-io/conan-center-index/blob/2ecc63a88002b1294f0ea0650f27c6012933aafa/recipes/libjpeg/all/conanfile.py#L106-L114